### PR TITLE
SONARPY-227 SONARPY-246 Fix import of coverage results

### PR DESCRIPTION
**Pull request by alban-auzeill**

### CODE REVIEW

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java [hunk 1]
```diff
@@ -59,99 +59,68 @@ public void execute(SensorContext context, Map<InputFile, Set<Integer>> linesOfC
...
-    Map<InputFile, NewCoverage> overallCoverageMeasures = parseReports(overallReports, context);
-    HashSet<InputFile> filesCoveredOverall = new HashSet<>();
-    saveMeasures(overallCoverageMeasures, filesCoveredOverall, CoverageType.OVERALL);
-
-    if (settings.getBoolean(FORCE_ZERO_COVERAGE_KEY)) {
+    HashSet<InputFile> filesCovered = new HashSet<>();
+    List<File> reports = new ArrayList<>();
+    reports.addAll(getReports(settings, baseDir, REPORT_PATH_KEY, DEFAULT_REPORT_PATH));
+    reports.addAll(getReports(settings, baseDir, IT_REPORT_PATH_KEY, DEFAULT_REPORT_PATH));
+    reports.addAll(getReports(settings, baseDir, OVERALL_REPORT_PATH_KEY, DEFAULT_REPORT_PATH));
```
Comment (pynicolas) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java]: Do we add `DEFAULT_REPORT_PATH` 3 times when no property is set?

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java [hunk 2]
```diff
@@ -59,99 +59,68 @@ public void execute(SensorContext context, Map<InputFile, Set<Integer>> linesOfC
...
+    if (!reports.isEmpty()) {
+      LOG.info("Python test coverage");
+      for (File report : reports) {
+        Map<InputFile, NewCoverage> coverageMeasures = parseReport(report, context);
+        saveMeasures(coverageMeasures, filesCovered);
+      }
+    }
+    if (settings.getBoolean(FORCE_ZERO_COVERAGE_KEY) || (!settings.hasKey(FORCE_ZERO_COVERAGE_KEY) && !reports.isEmpty())) {
```
Comment (pynicolas) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java]: Why do we add this new condition?
```(!settings.hasKey(FORCE_ZERO_COVERAGE_KEY) && !reports.isEmpty())```

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/CoberturaParser.java [hunk 3]
```diff
@@ -40,38 +40,50 @@ public void parseReport(File xmlFile, SensorContext context, final Map<InputFile
...
-    while (pack.getNext() != null) {
-      collectFileMeasures(pack.descendantElementCursor("class"), context, coverageData);
+  private static File extractBaseDirectory(SMInputCursor sources, File defaultBaseDirectory) throws XMLStreamException {
+    SMInputCursor source = sources.childElementCursor("source");
+    while (source.getNext() != null) {
+      String path = source.collectDescendantText();
+      if (!StringUtils.isBlank(path)) {
+        File baseDirectory = new File(path);
+        if (!baseDirectory.isDirectory()) {
+          throw new XMLStreamException("Invalid directory path in 'source' element: " + path);
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/CoberturaParser.java]: I changed to only log a warning.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/CoberturaParser.java [hunk 4]
```diff
@@ -40,38 +40,50 @@ public void parseReport(File xmlFile, SensorContext context, final Map<InputFile
...
     }
+    return defaultBaseDirectory;
   }
 
-  private static void collectFileMeasures(SMInputCursor classCursor, SensorContext context, Map<InputFile, NewCoverage> coverageData) throws XMLStreamException {
+  private static void collectFileMeasures(SMInputCursor classCursor, SensorContext context, Map<InputFile, NewCoverage> coverageData, File baseDirectory)
+    throws XMLStreamException {
     while (classCursor.getNext() != null) {
-      String fileName = classCursor.getAttrValue("filename");
-      InputFile inputFile = context.fileSystem().inputFile(context.fileSystem().predicates().hasPath(fileName));
```
[the remark this answers is not in the document]
Comment (alban-auzeill) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/CoberturaParser.java]: Good point, I added the support of absolute path.

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java (hunk 1 above)
Comment (alban-auzeill) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java]: I restored the right default values

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java (hunk 2 above)
Comment (alban-auzeill) [reply sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java]: Removed from here and https://jira.sonarsource.com/browse/SONARPY-246

File: sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java [hunk 5]
```diff
@@ -53,105 +53,73 @@
...
-        .ofType(ctype);
+        .onFile(inputFile);
       linesOfCode.forEach((Integer line) -> newCoverage.lineHits(line, 0));
       newCoverage.save();
     }
   }
 
-
-  private Map<InputFile, NewCoverage> parseReports(List<File> reports, SensorContext context) {
+  private Map<InputFile, NewCoverage> parseReport(File report, SensorContext context) {
```
Comment (SonarTech) [inline sonar-python-plugin/src/main/java/org/sonar/plugins/python/coverage/PythonCoverageSensor.java]: Code Smell Code Smell: Make "parseReport" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.python%3Apython&branch=SONARPY-227&issues=AWDq-kQSvOZGGKctVrGW&open=AWDq-kQSvOZGGKctVrGW)

[ISSUE_KEY]: AWDq-kQSvOZGGKctVrGW

**Pull request merged**